x-pack/plugins/lens/public/app_plugin/app.tsx

```diff
@@ -220,379 +198,410 @@ export function App({
           defaultMessage: 'Visualize',
         }),
-      },
-      {
-        text: state.persistedDoc
-          ? state.persistedDoc.title
-          : i18n.translate('xpack.lens.breadcrumbsCreate', { defaultMessage: 'Create' }),
-      },
-    ]);
+      });
+    }
+    let currentDocTitle = i18n.translate('xpack.lens.breadcrumbsCreate', {
+      defaultMessage: 'Create',
+    });
+    if (state.persistedDoc) {
+      currentDocTitle = byValueMode
+        ? i18n.translate('xpack.lens.breadcrumbsByValue', { defaultMessage: 'Edit Visualization' })
+        : state.persistedDoc.title;
+    }
+    breadcrumbs.push({ text: currentDocTitle });
+    chrome.setBreadcrumbs(breadcrumbs);
   }, [
-    core.application,
-    core.chrome,
-    core.http.basePath,
+    dashboardFeatureFlag.allowByValueEmbeddables,
+    state.isLinkedToOriginatingApp,
+    getOriginatingAppName,
     state.persistedDoc,
-    originatingApp,
-    redirectTo,
-    getAppNameFromId,
+    redirectToOrigin,
+    initialInput,
+    application,
+    chrome,
   ]);
 
-  useEffect(
-    () => {
-      if (docId && (!state.persistedDoc || state.persistedDoc.id !== docId)) {
-        setState((s) => ({ ...s, isLoading: true }));
-        docStorage
-          .load(docId)
-          .then((doc) => {
-            core.chrome.recentlyAccessed.add(getFullPath(docId), doc.title, docId);
-            getAllIndexPatterns(
-              _.uniq(
-                doc.references.filter(({ type }) => type === 'index-pattern').map(({ id }) => id)
-              ),
-              data.indexPatterns,
-              core.notifications
-            )
-              .then((indexPatterns) => {
-                // Don't overwrite any pinned filters
-                data.query.filterManager.setAppFilters(
-                  injectFilterReferences(doc.state.filters, doc.references)
-                );
-                setState((s) => ({
-                  ...s,
-                  isLoading: false,
-                  persistedDoc: doc,
-                  lastKnownDoc: doc,
-                  query: doc.state.query,
-                  indexPatternsForTopNav: indexPatterns,
-                }));
-              })
-              .catch((e) => {
-                setState((s) => ({ ...s, isLoading: false }));
+  useEffect(() => {
+    if (
+      !initialInput ||
+      (attributeService.inputIsRefType(initialInput) &&
+        initialInput.savedObjectId === state.persistedDoc?.savedObjectId)
+    ) {
+      return;
+    }
 
-                redirectTo();
-              });
+    setState((s) => ({ ...s, isLoading: true }));
+    attributeService
+      .unwrapAttributes(initialInput)
+      .then((attributes) => {
+        if (!initialInput) {
+          return;
+        }
+        const doc = {
+          ...initialInput,
+          ...attributes,
+          type: LENS_EMBEDDABLE_TYPE,
+        };
+
+        if (attributeService.inputIsRefType(initialInput)) {
+          chrome.recentlyAccessed.add(
```

Comment:
I didn't see any tests for this?

Reply:
There are two, but they weren't named properly. I have renamed them to `adds to the recently accessed list on load` and `adds to the recently accessed list on save`. You can find them under 'persistence'.